include/ginkgo/core/base/utils.hpp

```diff
@@ -325,6 +325,82 @@ inline const typename std::decay<T>::type *as(const U *obj)
 }
 
 
+/**
+ * Performs polymorphic type conversion of a unique_ptr.
+ *
+ * @tparam T  requested result type
+ * @tparam U  static type of the passed object
+ *
+ * @param obj  the unique_ptr to the object which should be converted.
+ *             If successful, it will be reset to a nullptr.
+ *
+ * @return If successful, returns a unique_ptr to the subtype, otherwise throws
+ *         NotSupported.
+ */
+template <typename T, typename U>
+inline std::unique_ptr<typename std::decay<T>::type> as(
```

Comment:
Somewhat related: Why do we need the `std::decay` here? T is always user-specified, so it should not be a reference in any case. Can we just drop it from `gko::as`?

Reply:
It also does `remove_cv` in particular. You do not want `const` to appear here so that seems fine to me.

Reply:
Alright, so I'll leave it as-is.